### PR TITLE
llvm-objdump/ELF: fix crash when reading dyn str table

### PR DESCRIPTION
**Pull request by artagnon**
When reading the dynamic string table, llvm-objdump used to crash if the ELF was malformed, due to an erroneous consumption of error status. Instead, propogate the error status to the caller, fixing the crash, and printing a warning.

### CODE REVIEW

**Review by jh7370 — COMMENTED** [review]
Thanks for the patch: the functional change looks fine, but I'd much prefer a more targeted and minimal test file. I'm assuming all that it needs is a single invalid DT_STRTAB and possibly one or two other tags. I'd also either merge the test into `dynamic-malformed.test` or rename it to something more descriptive, like `dynamic-invalid-addr.test`

**Comment by artagnon** [comment]
Thanks for the review @jh7370; now fixed.

File: llvm/test/tools/llvm-objdump/ELF/dynamic-malformed.test [hunk 1]
```diff
@@ -29,10 +28,36 @@ FileHeader:
...
   - Name: .dynamic
     Type: SHT_DYNAMIC
     Entries:
       - Tag:   DT_SONAME
         Value: 1
+
+# RUN: yaml2obj %s --docnum=3 -o %t.invalidaddr
+# RUN: llvm-objdump -p %t.invalidaddr | FileCheck %s --check-prefix=ADDR
+# RUN: llvm-objdump -p %t.invalidaddr 2>&1 >/dev/null | \
+# RUN: FileCheck %s -DFILE=%t.invalidaddr --implicit-check-not=warning --check-prefix=ADDR-ERROR
```
Comment (jh7370) [inline llvm/test/tools/llvm-objdump/ELF/dynamic-malformed.test:41]: Assuming what you're trying to achieve here is check stdout and stderr independently, I wouldn't bother: we usually just check them all in the combined output, so I'd write things like this:

```
# RUN: llvm-objdump -p %t.invalidaddr 2>&1 | \
# RUN:   FileCheck %s --check-prefix=ADDR --implicit-check-not-warning=warning:
```
and then just reorder the check lines accordingly. A couple of other minor notes from that by the way: a) I suggest including the colon after the implicit-check-not pattern, to reduce the risk of spurious matches against things like file paths; b) I find it more readable to break lines in the manner illustrated (`| \` at the end of the first line and additional spaces at the start of the second).

Comment (artagnon) [reply llvm/test/tools/llvm-objdump/ELF/dynamic-malformed.test]: Okay, the printing order is now fixed, but upon further investigation, it looks like the original bug is not fixed now.

File: llvm/tools/llvm-objdump/ELFDump.cpp [hunk 2]
```diff
@@ -232,14 +231,16 @@ template <class ELFT> void ELFDumper<ELFT>::printDynamicSection() {
         Dyn.d_tag == ELF::DT_AUXILIARY || Dyn.d_tag == ELF::DT_FILTER) {
       Expected<StringRef> StrTabOrErr = getDynamicStrTab(Elf);
       if (StrTabOrErr) {
-        const char *Data = StrTabOrErr.get().data();
-        outs() << (Data + Dyn.d_un.d_val) << "\n";
+        const char *Data = StrTabOrErr->data();
+        outs() << format(TagFmt.c_str(), Str.c_str());
+        outs() << Data + Dyn.getVal() << "\n";
```
Comment (MaskRay) [inline llvm/tools/llvm-objdump/ELFDump.cpp:236]: `<< format(...) << (Data + Dyn.getVal()) << '\n'`

ditto below

**Review by MaskRay — APPROVED** [review]

**Review by jh7370 — APPROVED** [review]
LGTM with @MaskRay's suggestion.

**Pull request merged**